### PR TITLE
Add the data getting part of tpcds connector

### PR DESCRIPTION
**Pull request by rschlussel-zz**
@petroav

### CODE REVIEW

File: presto-main/etc/catalog/tpcds.properties [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#
```
Comment (petroav) [inline presto-main/etc/catalog/tpcds.properties]: The existing .properties files don't have a license header like this

**Comment by rschlussel-zz** [comment]
hold off on review I've found a bug with selecting a subset of columns.

**Comment by rschlussel-zz** [comment]
fixed and updated.  The fix requires https://github.com/Teradata/tpcds/pull/5

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java [hunk 2]
```diff
@@ -263,7 +269,7 @@ private static ConnectorTableMetadata getTableMetadata(String schemaName, Table
         ImmutableList.Builder<ColumnMetadata> columns = ImmutableList.builder();
         for (Column column : tpcdsTable.getColumns()) {
             columns.add(new ColumnMetadata(column.getName(), getPrestoType(column.getType())));
-    }
+        }
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:272]: Accidental?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java [hunk 3]
```diff
@@ -156,10 +157,6 @@ else if (tableName.equals(Table.WEB_RETURNS.getName())) {
             partitioningColumn = Optional.of(columns.get(WebReturnsColumn.WR_ORDER_NUMBER.getName()));
             tpcdsTable = Optional.of(Table.WEB_RETURNS);
         }
-        else if (tableName.equals(Table.INVENTORY.getName())) {
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:159]: Why did you remove this?

EDIT: because INVENTORY doesn't have a partitioning column?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java [hunk 4]
```diff
@@ -229,11 +226,20 @@ else if (tableName.equals(Table.TIME_DIM.getName())) {
...
         }
 
-        if (!partitioningColumn.isPresent()) {
-            throw new IllegalArgumentException(format("Table %s does not exist", tableHandle.getTableName()));
+        ConnectorTableLayout layout;
+        if (partitioningColumn.isPresent()) {
+            layout = createLayoutForPartitioningColumn(tableHandle, partitioningColumn.get(), tpcdsTable.get());
+        }
+        else {
+            layout = new ConnectorTableLayout(
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:234]: Is this for the Inventory layout? Maybe add a comment explaining since the next ready won't have the context that Inventory was removed from the massive if else block.

Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:234]: and the version table

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java (hunk 3 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:159]: exactly

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+// Right now, splits are just the entire TPCH table
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java]: Hehe, the copy pasta got you as well :)

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        public TpcdsRecordCursor(Iterator<List<List<String>>> rows, List<Column> columns)
+        {
+            this.rows = rows;
+            this.columns = columns;
+        }
+
+        @Override
+        public long getTotalBytes()
+        {
+            return 0;
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:89]: What are these methods for, do you know?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+                row = null;
+                return false;
+            }
+
+            row = rows.next().get(0);
+            return true;
+        }
+
+        @Override
+        public boolean getBoolean(int field)
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:124]: Do you know what this method is?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 6 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:89]: no idea

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 7 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:124]: yeah if the java type for your column was boolean this would get the value from the results and return the appropriate boolean.  TPC-DS doesn't have any boolean columns, though

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java (hunk 5 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java]: copy pasta is my favorite kind of pasta :)

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java (hunk 2 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:272]: No, it was insufficiently indented before.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 6 above)
Comment (cawallin) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:89]: It's for Presto to be able to report stats better (see the javadoc on RecordCursor). getTotalBytes() is the total number of bytes you'll ever have (you may or may not have this info so if you don't, it's 0). getCompletedBytes() is the number of bytes you've gone through so far. getReadTimeNanos() isn't necessary

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        @Override
+        public boolean advanceNextPosition()
+        {
+            if (closed || !rows.hasNext()) {
+                closed = true;
+                row = null;
+                return false;
+            }
+
+            row = rows.next().get(0);
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:119]: So given that parent and child are generated in tuples, does this line mean that we're effectively ignoring the child?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 7 above)
Comment (petroav) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:124]: Oh wow, that's surprising (that TPC-DS doesn't have any boolean types).

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 8 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:119]: well we don't bother generating the child at all if we're only generating one table, as we're doing here.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 7 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:124]: Yeah, everything that would be a boolean becomes a String ("Y" or "N")

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public long getLong(int field)
+        {
+            checkState(row != null, "No current row");
+            Column column = columns.get(field);
+            if (column.getType().getBase() == ColumnType.Base.DATE) {
+                return LocalDate.parse(row.get(column.getPosition())).toEpochDay();
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:135]: Is `java.time` the right package to use? Does Presto use that or does it use Joda? @cawallin @maciejgrzybek

Comment (cawallin) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:135]: Presto primarily uses Joda, from what I see in the code

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+            }
+            if (column.getType().getBase() == ColumnType.Base.TIME) {
+                return LocalTime.parse(row.get(column.getPosition())).toNanoOfDay() / 1000000;
+            }
+            if (column.getType().getBase() == ColumnType.Base.INTEGER) {
+                return parseInt(row.get(column.getPosition()));
+            }
+            if (column.getType().getBase() == ColumnType.Base.DECIMAL) {
+                DecimalParseResult decimalParseResult = Decimals.parse(row.get(column.getPosition()));
+                return rescale((Long) decimalParseResult.getObject(), decimalParseResult.getType().getScale(), ((DecimalType) columnTypes.get(field)).getScale());
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:145]: Why do we need to rescale? Wouldn't `Decimals.parse()` return the correct scale?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 7 above)
Comment (maciejgrzybek) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:124]: Oracle wouldn't pass TPC-DS if they were any bools. ;)

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+        @Override
+        public Slice getSlice(int field)
+        {
+            checkState(row != null, "No current row");
+            Column column = columns.get(field);
+            if (column.getType().getBase() == ColumnType.Base.DECIMAL) {
+                return (Slice) Decimals.parse(row.get(column.getPosition())).getObject();
+            }
+            return Slices.utf8Slice(row.get(columns.get(field).getPosition()));
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:165]: I don't know if this is the right way to parse Slices. @cawallin @maciejgrzybek

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 9 above)
Comment (maciejgrzybek) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:135]: Airbase provides Joda, thus it should be preferred I'd say.

Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:135]: really?  joda tells you that starting in java8 you should use java.time.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 11 above)
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java]: Looks good!

Comment (maciejgrzybek) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:165]: Assuming it's UTF-8 encoded (and we have such an assumption in Presto for strings) it's okay.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 9 above)
Comment (cawallin) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:135]: I think I recall some other discussion about this, I think FB probably just hasn't migrated everything to java.time yet.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSetProvider.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+            int partNumber,
+            int totalParts)
+    {
+        ImmutableList.Builder<Column> builder = ImmutableList.builder();
+        for (ColumnHandle column : columns) {
+            String columnName = checkType(column, TpcdsColumnHandle.class, "column").getColumnName();
+            builder.add(table.getColumn(columnName));
+        }
+
+        Session session = Session.getDefaultSession().withScale(scaleFactor).withParallelism(totalParts).withChunkNumber(partNumber + 1).withTable(table);
```
Comment (petroav) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSetProvider.java]: I think this line is more readable if you break it up like so:

```
        Session session = Session.getDefaultSession()
                .withScale(scaleFactor)
                .withParallelism(totalParts)
                .withChunkNumber(partNumber + 1)
                .withTable(table);
```

Don't feel too strongly about this, so it's totally up to you.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 10 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:145]: No, it needs to match the scale of the column type.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 11 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:165]: it's what everything else uses to get Slices from strings and is what is used in the tpch connector.

**Review by petroav — COMMENTED** [review]
A couple minor questions/comments and also soliciting help from @cawallin and @maciejgrzybek for a couple of the lines.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 9 above)
Comment (petroav) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:135]: Maybe we should ask on the #general Slack channel?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 11 above)
Comment (maciejgrzybek) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:165]: Yup, it's okay :)

Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:165]: Although actually I'm remembering tpcds uses latin1 encoding (Relevant for one particular field that sometimes has special characters).  I'll deal with translating it on the generator side

Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:165]: nevermind.. doesn't matter

**Review by petroav — APPROVED** [review]
Looks good. I would rename the commit message to `Add the data generation portion of the TPC-DS connector`

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+public class TpcdsBucketFunction
+        implements BucketFunction
+{
+    private final int bucketCount;
+    private final long rowsPerBucket;
+
+    public TpcdsBucketFunction(int bucketCount, long rowsPerBucket)
+    {
+        this.bucketCount = bucketCount;
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java:35]: `checkArgument(bucketCount > 1)`

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+        this.rowsPerBucket = rowsPerBucket;
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java:36]: `checkArgument(rowsPerBucket > 1)`

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public int getBucket(Page page, int position)
+    {
+        Block block = page.getBlock(0);
+        if (block.isNull(position)) {
+            return 0;
+        }
+
+        long rowNumber = BIGINT.getLong(block, position);
+        int bucket = Ints.checkedCast(rowNumber / rowsPerBucket);
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java:48]: wrap with`try-catch` and throw proper `PrestoException` instead of generic `ArithmeticException`.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java (hunk 4 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:234]: done

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java [hunk 16]
```diff
@@ -229,11 +226,20 @@ else if (tableName.equals(Table.TIME_DIM.getName())) {
...
+                    new TpcdsTableLayoutHandle(tableHandle),
+                    Optional.empty(),
+                    TupleDomain.all(),
+                    Optional.empty(),
+                    Optional.empty(),
+                    Optional.empty(),
+                    new ArrayList<>());
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java]: `ImmutableList.of()`

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+
+    private final List<Type> columnTypes;
+
+    public TpcdsRecordSet(Results results, List<Column> columns)
+    {
+        requireNonNull(results, "results is null");
+
+        this.results = results;
+        this.columns = ImmutableList.copyOf(columns);
+        this.columnTypes = ImmutableList.copyOf(this.columns.stream().map(column -> getPrestoType(column.getType())).collect(Collectors.toList()));
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java]: `ImmutableCollectors` instead of `ImmutableList.copyOf(stream.collect(toList()))`

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+            implements RecordCursor
+    {
+        private final Iterator<List<List<String>>> rows;
+        private final List<Column> columns;
+        private List<String> row;
+        private boolean closed;
+
+        public TpcdsRecordCursor(Iterator<List<List<String>>> rows, List<Column> columns)
+        {
+            this.rows = rows;
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java]: `requireNonNull`

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+            this.columns = columns;
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java]: ditto

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+    public class TpcdsRecordCursor
+            implements RecordCursor
+    {
+        private final Iterator<List<List<String>>> rows;
+        private final List<Column> columns;
+        private List<String> row;
+        private boolean closed;
+
+        public TpcdsRecordCursor(Iterator<List<List<String>>> rows, List<Column> columns)
+        {
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:81]: Initialize `this.closed`

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+    private final String connectorId;
+    private final NodeManager nodeManager;
+    private final int splitsPerNode;
+
+    public TpcdsNodePartitioningProvider(String connectorId, NodeManager nodeManager, int splitsPerNode)
+    {
+        checkArgument(splitsPerNode > 0, "splitsPerNode must be at least 1");
+
+        this.connectorId = connectorId;
+        this.nodeManager = nodeManager;
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java:53]: `requireNonNull`

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+    public ToIntFunction<ConnectorSplit> getSplitBucketFunction(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
+    {
+        return value -> checkType(value, TpcdsSplit.class, "value").getPartNumber();
+    }
+
+    @Override
+    public BucketFunction getBucketFunction(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle, List<Type> partitionChannelTypes, int bucketCount)
+    {
+        long totalRows = checkType(partitioningHandle, TpcdsPartitioningHandle.class, "functionHandle").getTotalRows();
+        long rowsPerBucket = totalRows / bucketCount;
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java:85]: How you handle reminder?
E.g. 10 rows in 3 buckets. 1 remaining won't be assigned

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 9 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:135]: leaving as is.  After looking into it a bit more, this ends up being simpler because java.time has a toEpochDays() method that joda does not have.  David only cared about what made sense locally.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSetProvider.java [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+
+public class TpcdsRecordSetProvider
+        implements ConnectorRecordSetProvider
+{
+    @Override
+    public RecordSet getRecordSet(ConnectorTransactionHandle transaction, ConnectorSession session, ConnectorSplit split, List<? extends ColumnHandle> columns)
+    {
+        TpcdsSplit tpcdsSplit = checkType(split, TpcdsSplit.class, "split");
+        String tableName = tpcdsSplit.getTableHandle().getTableName();
+        Table table = getTable((tableName));
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSetProvider.java]: double parenthesis?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSetProvider.java (hunk 12 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSetProvider.java]: done

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+            @JsonProperty("addresses") List<HostAddress> addresses)
+    {
+        checkState(partNumber >= 0, "partNumber must be >= 0");
+        checkState(totalParts >= 1, "totalParts must be >= 1");
+        checkState(totalParts > partNumber, "totalParts must be > partNumber");
+
+        this.tableHandle = requireNonNull(tableHandle, "tableHandle is null");
+        this.partNumber = partNumber;
+        this.totalParts = totalParts;
+        this.addresses = ImmutableList.copyOf(requireNonNull(addresses, "addresses is null"));
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java]: `requireNonNull` is not needed if using `ImmutableList.copyOf`. Won't harm though.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplitManager.java [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+        implements ConnectorSplitManager
+{
+    private final String connectorId;
+    private final NodeManager nodeManager;
+    private final int splitsPerNode;
+
+    public TpcdsSplitManager(String connectorId, NodeManager nodeManager, int splitsPerNode)
+    {
+        this.connectorId = connectorId;
+        this.nodeManager = nodeManager;
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplitManager.java:48]: `requireNonNull`

**Review by maciejgrzybek — CHANGES_REQUESTED** [review]
Minor comments + the one with `getBucketFunction`.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java (hunk 15 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java:48]: This is copied directly from TpchBucketFunction.  Should I change it anyway?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java (hunk 16 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java]: done

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 17 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java]: done

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 20 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:81]: booleans by default are initialized to false.  Is that a style we're consistent about?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 18 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java]: done

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java (hunk 19 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java]: done

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java (hunk 15 above)
Comment (maciejgrzybek) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java:48]: It's rather unusual case to make `checkedCast` throw here but if we can make error more readable to the user, then why not?

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java (hunk 21 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java:53]: done

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java (hunk 22 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java:85]: the remainder all goes in the last bucket (see the bucket function)

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSetProvider.java (hunk 23 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSetProvider.java]: fixed :)

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java (hunk 24 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java]: This was also copy paste from the tpcds connector :).  But i'll get rid of it

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplitManager.java (hunk 25 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplitManager.java:48]: done

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java (hunk 13 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java:35]: done (>= 1)

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java (hunk 14 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java:36]: done (>=1)

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java (hunk 15 above)
Comment (rschlussel-zz) [reply presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsBucketFunction.java:48]: leaving as is.  Looking at it again, we define the number of rows per bucket as totalRows/ nubmerOfBuckets, so it should be possible to run into this.  This is more like a checkArgument.

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+public class TpcdsNodePartitioningProvider
+        implements ConnectorNodePartitioningProvider
+{
+    private final String connectorId;
+    private final NodeManager nodeManager;
+    private final int splitsPerNode;
+
+    public TpcdsNodePartitioningProvider(String connectorId, NodeManager nodeManager, int splitsPerNode)
+    {
+        requireNonNull(connectorId);
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java]: message missing

File: presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+        requireNonNull(nodeManager);
```
Comment (maciejgrzybek) [inline presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java]: ditto

**Review by maciejgrzybek — APPROVED** [review]
Looks good
